merkledag/merkledag.go

```diff
@@ -394,56 +394,54 @@ func EnumerateChildren(ctx context.Context, ds LinkService, root *cid.Cid, visit
 var FetchGraphConcurrency = 8
 
 func EnumerateChildrenAsync(ctx context.Context, ds DAGService, c *cid.Cid, visit func(*cid.Cid) bool) error {
-	if !visit(c) {
-		return nil
-	}
-
-	root, err := ds.Get(ctx, c)
-	if err != nil {
-		return err
-	}
-
-	feed := make(chan node.Node)
-	out := make(chan *NodeOption)
+	feed := make(chan *cid.Cid)
+	out := make(chan []*cid.Cid)
 	done := make(chan struct{})
 
 	var setlk sync.Mutex
-
+	
+	errChan := make(chan error)
+	fetchersCtx, cancel := context.WithCancel(ctx)
+	
 	for i := 0; i < FetchGraphConcurrency; i++ {
 		go func() {
-			for n := range feed {
-				links := n.Links()
-				cids := make([]*cid.Cid, 0, len(links))
-				for _, l := range links {
-					setlk.Lock()
-					unseen := visit(l.Cid)
-					setlk.Unlock()
-					if unseen {
-						cids = append(cids, l.Cid)
-					}
+			for ic := range feed {
+				n, err := ds.Get(ctx, ic)
+				if err != nil {
+					errChan <- err
+					cancel()
+					return
 				}
-
-				for nopt := range ds.GetMany(ctx, cids) {
+				
+				setlk.Lock()
+				unseen := visit(ic)
+				setlk.Unlock()
+				
+				if unseen {
+					cids := make([]*cid.Cid, len(n.Links()))
+					for i, l := range n.Links() {
```

Comment:
my worry is that its an extra allocation per node that we could avoid by simply moving this for loop down into the processing below.